crates/payload/basic/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+impl Default for BasicPayloadJobGeneratorConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(1),
+            // 12s slot time
+            deadline: Duration::from_secs(12),
+            max_payload_tasks: 3,
+        }
+    }
+}
+
+/// A basic payload job that continuously builds a payload with the best transactions from the pool.
+pub struct BasicPayloadJob<Client, Pool, Tasks> {
+    /// The configuration for how the payload will be created.
+    config: PayloadConfig,
+    /// The client that can interact with the chain.
+    client: Client,
+    /// The transaction pool.
+    pool: Pool,
+    /// How to spawn building tasks
+    executor: Tasks,
+    /// The deadline when this job should resolve.
+    deadline: Pin<Box<Sleep>>,
+    /// The interval at which the job should build a new payload after the last.
+    interval: Interval,
+    /// The best payload so far.
+    best_payload: Arc<BuiltPayload>,
+    /// Receiver for the block that is currently being built.
+    pending_block: Option<PendingPayload>,
+    /// Restricts how many generator tasks can be executed at once.
+    payload_task_guard: PayloadTaskGuard,
+}
+
+impl<Client, Pool, Tasks> Stream for BasicPayloadJob<Client, Pool, Tasks>
+where
+    Client: StateProviderFactory + Clone + Unpin + 'static,
+    Pool: TransactionPool + Unpin + 'static,
+    Tasks: TaskSpawner + Clone + 'static,
+{
+    type Item = Result<Arc<BuiltPayload>, PayloadBuilderError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        // check if the deadline is reached
+        let deadline_reached = this.deadline.as_mut().poll(cx).is_ready();
+
+        // check if the interval is reached
+        if this.interval.poll_tick(cx).is_ready() &&
+            this.pending_block.is_none() &&
+            !deadline_reached
+        {
+            trace!("spawn new payload build task");
+            let (tx, rx) = oneshot::channel();
+            let client = this.client.clone();
+            let pool = this.pool.clone();
+            let cancel = Cancelled::default();
+            let _cancel = cancel.clone();
+            let guard = this.payload_task_guard.clone();
+            let payload_config = this.config.clone();
+            this.executor.spawn_blocking(Box::pin(async move {
+                // acquire the permit for executing the task
+                let _permit = guard.0.acquire().await;
+                build_payload(client, pool, payload_config, cancel, tx)
+            }));
+            this.pending_block = Some(PendingPayload { _cancel, payload: rx });
+        }
+
+        // poll the pending block
+        if let Some(mut fut) = this.pending_block.take() {
+            match fut.poll_unpin(cx) {
+                Poll::Ready(Ok(payload)) => {
+                    this.interval.reset();
+                    // TODO check if payload is better
```

Comment:
would we just compare the `BuiltPayload` fees here? Is there anything else involved with checking that the payload is better?

Reply:
by default no, it's just:

https://github.com/ethereum/go-ethereum/blob/4a9fa31450d3cdcea84735b68cd5a0a8450473f8/miner/payload_building.go#L96-L99